codec/proto: reuse of marshal byte buffers

adtac opened the pull request:
Performance benchmarks can be found below. Obviously, a 8 KiB
request/response is tailored to showcase this improvement as this is
where codec buffer reuse shines, but I've run other benchmarks too (like
1-byte requests and responses) and there's no discernable impact on
performance.

We do not allow reuse of buffers when stat handlers or binlogs are
turned on. This is because those two may need access to the data and
payload even after the data has been written to the wire. In such cases,
we never return the data back to the pool.

A buffer reuse threshold of 1 KiB was determined after several
experiments. There's diminished returns when buffer reuse is enabled for
smaller messages (actually, a negative impact).

```
unary-networkMode_none-bufConn_false-keepalive_false-benchTime_40s-trace_false-latency_0s-kbps_0-MTU_0-maxConcurrentCalls_6-reqSize_8192B-respSize_8192B-compressor_off-channelz_false-preloader_false
               Title       Before        After Percentage
            TotalOps       839638       906223     7.93%
             SendOps            0            0      NaN%
             RecvOps            0            0      NaN%
            Bytes/op    103788.29     80592.47   -22.35%
           Allocs/op       183.33       189.30     3.27%
             ReqT/op 1375662899.20 1484755763.20     7.93%
            RespT/op 1375662899.20 1484755763.20     7.93%
            50th-Lat    238.746µs    225.019µs    -5.75%
            90th-Lat    514.253µs    456.439µs   -11.24%
            99th-Lat    711.083µs    702.466µs    -1.21%
             Avg-Lat     285.45µs    264.456µs    -7.35%
```

Closes https://github.com/grpc/grpc-go/issues/2816

adtac commented on stream.go:
```diff
@@ -1165,8 +1182,8 @@ func (as *addrConnStream) SendMsg(m interface{}) (err error) {
-	// load hdr, payload, data
-	hdr, payld, _, err := prepareMsg(m, as.codec, as.cp, as.comp)
+	// load hdr, payload, data, returnBuffer
+	hdr, payld, _, returnBuffer, err := prepareMsg(m, as.codec, as.cp, as.comp)
```
Correct me if I'm wrong, but addrConnStream does not do any stats/binlog stuff?

adtac commented:
Well this turned out to be much more complicated than I originally though:

 - We can't have buffer reuse when stats handlers and binlogs are enabled because these can store references to the buffer internally.
 - Buffer re-use is tricky when client retries are involved.

Previously, I returned the buffer back to the codec layer as soon as the transport layer (read: loopy) was done with the buffer, but that's not a good with retries because the buffered slice of retry funcs will retain references to the payload, which may actually be the plain encoded data if no compression is used. Therefore, we may return the buffer back to the encoder only after the entire RPC is concluded, so we need to store all the buffers to return.

adtac commented:
Why is `bufferReuseThreshold = 1024`? It's because everything below it negatively impacts performance. Anything higher than that wastes memory when buffer reuse could positively impact performance.

![qps](https://user-images.githubusercontent.com/7521600/68816047-1b0a1b00-0632-11ea-8f65-f5daf823845f.png)
![bytes](https://user-images.githubusercontent.com/7521600/68816048-1cd3de80-0632-11ea-9555-4d7e63dc4bc3.png)
![50th](https://user-images.githubusercontent.com/7521600/68816051-1e9da200-0632-11ea-86a9-84817c8ebcd7.png)
![90th](https://user-images.githubusercontent.com/7521600/68816055-1fcecf00-0632-11ea-9fa2-46930aa29839.png)

adtac commented on encoding/encoding.go:
```diff
@@ -86,6 +86,24 @@ type Codec interface {
+// re-use using this method.
+//
+// This API is EXPERIMENTAL.
+type BufferedCodec interface {
```
Not sure if this is the best name for this; open to suggestions

adtac replied:
Done.

dfawley commented on vet.sh:
```diff
@@ -119,7 +119,7 @@ fi
 staticcheck -go 1.9 -checks 'inherit,-ST1015' ./... > "${SC_OUT}" || true
 # Error if anything other than deprecation warnings are printed.
-(! grep -v "is deprecated:.*SA1019" "${SC_OUT}")
+(! grep -Pv "\((SA1019)|(SA6002)\)$" "${SC_OUT}")
```
What's 6002?

We can ignore this, if it's appropriate to, by changing the command above to include "-SA6002" instead.

adtac replied:
SA6002 complains about the use of `sync.Pool` with a non-struct type (i.e. `[]byte`). It wants us to use a `bytes.Buffer`, but we can't change that now...

adtac replied:
Also done.

adtac commented on codec.go:
```diff
@@ -31,6 +31,24 @@ type baseCodec interface {
+// the returned byte buffer, they can choose to return it back to the encoding
+// library for re-use using this method.
+//
+// This API is EXPERIMENTAL.
```
Done. Also renamed this to `reusableBaseCodec`.

adtac commented on encoding/proto/proto.go:
```diff
@@ -88,23 +74,39 @@ func (codec) Unmarshal(data []byte, v interface{}) error {
+	// Make sure we set the length of the buffer to zero so that future appends
+	// will start from the zeroeth byte, not append to the previous, stale data.
+	//
+	// Apparently, sync.Pool with non-pointer objects (slices, in this case)
```
Then we'd be doing an allocation are put-time. Get would return a reference to a `*[]byte`, which we'd need to dereference to get the underlying byte slice when returning from `Marshal` (can't change the `Marshal` API at this point). When `ReturnBuffer` is called, we will need to create a different reference to the same byte slice. Escape analysis would point that the new `*[]byte` escapes `ReturnBuffer`, so it'd be allocated on the heap. This just moves the problem.

The only way to _really_ solve this is to use `bytes.Buffer` as the return value from `Marshal`. Then, we'd be able to pass the same `bytes.Buffer` to `ReturnBuffer` -- zero allocations. (I suppose `bytes.Buffer` is basically a `*[]byte`.)

adtac commented on preloader.go:
```diff
@@ -55,6 +56,14 @@ func (p *PreparedMsg) Encode(s Stream, msg interface{}) error {
 	p.encodedData = data
+	if len(data) >= bufferReuseThreshold {
+		if bcodec, ok := rpcInfo.preloaderInfo.codec.(bufferedBaseCodec); ok {
+			p.returnBuffer = func() {
```
Yes. I don't use a preloader, but I imagine it exists for a reason. I traced the history behind this and it looks like it was created for applications that want encoding to be separate from transport-related operations. In that case, it will be invoked when the application calls `prepareMsg`, which uses `p.returnBuffer` if the message is a `*PreparedMsg`.

dfawley replied:
I don't see where `PreparedMsg.returnBuffer` is ever called, though(?).

adtac replied:
The `returnBuffer` value returned in `prepareMsg` is used by `SendMsg`. If the user passes a `*PreparedMsg` to `SendMsg`, this function would eventually be called by the `Done()` call in loopy.

adtac commented on server.go:
```diff
@@ -1039,7 +1051,15 @@ func (s *Server) processUnaryRPC(t transport.ServerTransport, stream *transport.
+		defer func() {
+			// Do not wait for the unary response to be written to the wire. Spawn a
+			// goroutine to wait on the waitgroup instead to be non-blocking.
+			go waitCallReturnBuffer(f, opts.ReturnBufferWaitGroup)
```
When `Done()` is called by loopy, it doesn't mean that we can free the buffer. On the client side, if a future message fails, every message needs to be retried -- even ones that were successfully written to the wire. So we'll need to keep an unreturned reference to these payloads so that they aren't recycled (which will corrupt the data).

For this, we need to wait till the RPC is committed so that it won't be retried. But that alone doesn't guarantee that these byte slices can be returned -- what if there are messages in loopy's `cbuf` that haven't been written to the wire? IIUC, a query is defined to be committed if it receives headers back from the server; in a bidirectional stream, a client may receive headers from the server before all client streaming messages are written out. We need to wait till both of these conditions are satisfied:

1. The RPC is committed (i.e. no more retries) -- which means we can't call `f()` immediately within `Done()` even if the counter is 0 because it may be retried in the future.
2. All outgoing messages have been written to the wire (i.e. no more references to the payload).

The issue now is that this may be satisfied after the end of `processUnaryRPC` if loopy has a large enough backlog to process.

adtac replied:
And AFAICT the only way to be able to do stuff well after the lifetime of an RPC call is by forking a goroutine.

dfawley replied:
Can we make it so that `cs.commitAttempt` and loopy finishing writing each do a `Done`?  `Add(2)` first, and when the both conditions are satisfied, we will return the buffer?

adtac replied:
Nice, I think that would work. It'd need to be a `Add(1)` for each attempt + `Add(1)` for loopy enqueue, but I see what you mean. Thanks, I'll try this!

dfawley replied:
When we commit the attempt, we never retry, so if we Done() in `commitAttempt` I think we should be good.  (I'm not sure if we call `commitAttempt` unconditionally; if not, we'd want to `Add(1)` instead of `2` if this RPC will not be retried.)

adtac replied:
Done. See PATCH V3.

adtac replied:
one sec, need to fix something, WIP for a few minutes, sorry

adtac replied:
@dfawley ok, this is once again ready for review.

adtac commented on internal/transport/controlbuf.go:
```diff
@@ -726,6 +727,16 @@ func (l *loopyWriter) outFlowControlSizeRequestHandler(o *outFlowControlSizeRequ
+		// call wg.Done() anymore.
+		if str.state == active {
+			for !str.itl.isEmpty() {
+				if dataItem, ok := str.itl.dequeue().(*dataFrame); ok && dataItem.wg != nil {
```
Good catch. Only issue I see with that is that this would require a interface type assertion for every dequeue to check if it's a `*dataFrame` even when we don't expect it to be. For example, `originateStream` expects a `*headerFrame` in the item list head.

This would also mix the item list linked list abstraction with gRPC-specific stuff (right now it's just a `interface{}` LL), but that's not really a problem since this is all internal stuff (but I thought I should mention it).

dfawley replied:
One idea to work around the abstraction leakage and type assertion: add an `onDequeue` field to `itemNode` that `itemList.dequeue()` calls unconditionally (if non-nil).  I don't think this would add too much overhead.

adtac replied:
Done.

adtac commented on internal/leakcheck/leakcheck.go:
```diff
@@ -42,6 +43,8 @@ var goroutinesToIgnore = []string{
 	"runtime_mcall",
 	"(*loggingT).flushDaemon",
 	"goroutine in C code",
+	"grpc.callReturnBuffers",
```
Yes. Some end2end tests check that there are no leaked goroutines immediately after a query is done. Specifically, `TestClientDoesntDeadlockWhileWritingErrornousLargeMessages` (nice name btw) creates 1,000 queries concurrently, so there's no guarantee that all 1,000 forked goroutines will be done. (The test sends 1MB requests, so it's well above the threshold for when buffer reuse is turned on.)

Perhaps I could add a new test to make sure the callReturnBuffers isn't actually leaked?

dfawley replied:
This shouldn't be necessary once the WaitGroup+goroutine is removed, though, right?

adtac replied:
Indeed, removing.

adtac replied:
Done.

I propose we still keep the leakcheck commit that uses a resizing buffer like runtime/debug. That'll still be useful in the future even though this PR does not actually change the behaviour of leakcheck.

adtac commented on preloader.go:
```diff
@@ -55,6 +56,14 @@ func (p *PreparedMsg) Encode(s Stream, msg interface{}) error {
 		return err
 	}
 	p.encodedData = data
+	if len(data) >= bufferReuseThreshold {
```
Done.

adtac commented on stream.go:
```diff
@@ -1507,23 +1598,44 @@ func MethodFromServerStream(stream ServerStream) (string, bool) {
+		return nil, nil, nil, nil, err
 	}
+
+	if attemptBufferReuse && len(data) >= bufferReuseThreshold {
```
Done.

dfawley commented on stream.go:
```diff
@@ -448,8 +458,14 @@ type csAttempt struct {
+		for _, rb := range cs.returnBuffers {
+			rb.Done()
+		}
+		cs.returnBuffers = nil
```
Since we are setting this to `nil`, do we need the `if !committed` at the top?

adtac replied:
Correct, done.

adtac commented on internal/transport/transport.go:
```diff
@@ -805,3 +808,44 @@ func ContextErr(err error) error {
+		// Same behaviour as sync.WaitGroup, this should NEVER happen. And if it
+		// does happen, it's better to terminate early than silently continue with
+		// corrupt data.
+		panic("grpc: ReturnBuffer negative counter")
```
Done.

adtac commented on internal/transport/transport.go:
```diff
@@ -805,3 +808,44 @@ func ContextErr(err error) error {
+// Each Add call must be paired with a corresponding Done call, which executes
+// the closured function (if any).
+func (rb *ReturnBuffer) Done() {
+	if rb.f == nil {
```
Done.

adtac commented:
@dfawley thanks for the review, I've addressed all comments. CI should pass this time. PTAL; if you'd like to merge this in the future, let me know and I'll squash everything into two logical commits -- one for the leaktest change, one for the actual reuse change.

dfawley commented:
FYI: travis is showing a data race that looks like it's caused by this PR.  Can you take a look?

adtac commented:
I literally can't reproduce this locally. I've run the failing test 8000 times (yes, 8000) with the race detector turned on and every single one test passed. And now Travis is green too. I don't want to say it's a memory corruption bug, but...

dfawley commented on codec.go:
```diff
@@ -31,6 +31,22 @@ type baseCodec interface {
+// the returned byte buffer, they can choose to return it back to the encoding
+// library for re-use using this method.
+type reusableBaseCodec interface {
+	baseCodec
```
This is unnecessary since you never use the baseCodec functions inside the type assertions checking for the `ReturnBuffer` method.  Delete and rename to `bufferReturner`?

adtac replied:
Done.

dfawley commented:
> I literally can't reproduce this locally. I've run the failing test 8000 times (yes, 8000) with the race detector turned on and every single one test passed. And now Travis is green too. I don't want to say it's a memory corruption bug, but...

travis? :man_shrugging:

tmthrgd commented:
Using a `sync.Pool` in this way is almost certainly incorrect and liable to cause performance problems and excessive heap growth: golang/go#23199 and golang/go#27735.
